File: C++/findUniqueNumberOfArray.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
    int bitmap[32];
    int count, value;
    int result = 0;
    scanf("%d", &count);
    for (int i = 0; i < count; i++)
    {
        scanf("%d", &value);
        int index = 31;
        while (value)
        {
            bitmap[index--] += value & 1;
            value = value >> 1;
        }
    }
    for (int i = 0; i < 32; i++)
        result = result << 1 + bitmap[i] % 3;
    printf("%d\n", result);
}

File: LeetCode/LeetCode-Offer/findUniqueNumberOfArray.php
<?php

/**
 * 题目描述
 * 输入一组未排序的整数，其中一个数字只出现一次，剩下的所有数字都出现了三次。
 * 找出这个只出现一次的数字。
 * 例如输入: [1,2,2,2,3,3,3]，输出为1
 */
function findUniqueNumberOfArray($array)
{
    $result = 0;
    $bitmap = array_fill(0, 32, 0);
    foreach ($array as $value) {
        $index = 31;
        while ($value) {
            $bitmap[$index--] += $value & 1;
            $value = $value >> 1;
        }
    }
    foreach ($bitmap as $value)
        $result = ($result << 1) + $value % 3;
    return $result;
}



// TODO:Test
$array = [1, 2, 2, 2, 3, 3, 3];
echo findUniqueNumberOfArray($array);
